discovery-provider/src/tasks/update_track_is_available.py

```diff
@@ -0,0 +1,250 @@
+    """
+    track_ids_and_replica_sets = (
+        session.query(Track.track_id, User.primary_id, User.secondary_ids)
+        .join(User, Track.owner_id == User.user_id, isouter=True)  # left join
+        .filter(
+            User.is_current == True,
+            Track.is_current == True,
+            Track.track_id.in_(track_ids),
+        )
+        .all()
+    )
+
+    return track_ids_and_replica_sets
+
+
+def query_tracks_by_track_ids(session: Any, track_ids: List[int]) -> List[Any]:
+    """Returns a list of Track objects that has a track id in `track_ids`"""
+    tracks = (
+        session.query(Track)
+        .filter(
+            Track.is_current == True,
+            Track.track_id.in_(track_ids),
+        )
+        .all()
+    )
+
+    return tracks
+
+
+def query_registered_content_node_info(
+    session: Any,
+    # ) -> List[Dict[str, Union[str, int]]]:
+) -> List[ContentNodeInfo]:
+    """Returns a list of all registered Content Node endpoint and spID"""
+    registered_content_nodes = (
+        session.query(URSMContentNode.endpoint, URSMContentNode.cnode_sp_id)
+        .filter(
+            URSMContentNode.is_current == True,
+        )
+        .all()
+    )
+
+    def create_named_fields(node):
+        return {"endpoint": node[0], "spID": node[1]}
+
+    return list(map(create_named_fields, registered_content_nodes))
+
+
+def check_track_is_available(
+    redis: Any, track_id: int, spID_replica_set: List[int]
+) -> bool:
+    """
+    Checks if a track is available in the replica set. Needs to only be available
+    on one replica set node to be marked as available.
+        redis: redis instance
+        track_id: the observed track id
+        spID_replica_set: an array of the SP IDs that are associated with track
+    """
+
+    i = 0
+    while i < len(spID_replica_set):
+        spID_unavailable_tracks_key = get_unavailable_tracks_redis_key(
+            spID_replica_set[i]
+        )
+        is_available_on_sp = not redis.sismember(spID_unavailable_tracks_key, track_id)
+
+        if is_available_on_sp:
+            return True
+
+        i = i + 1
+
+    return False
+
+
+def get_unavailable_tracks_redis_key(spID: int) -> str:
+    """Returns the redis key used to store the unavailable tracks on a sp"""
+    return f"update_track_is_available:unavailable_tracks_{spID}"
+
+
+# ####### CELERY TASKS ####### #
+@celery.task(name="update_track_is_available", bind=True)
+def update_track_is_available(self) -> None:
+    """
+    Recurring task that updates whether tracks are available on the network
+    """
+
+    db = update_track_is_available.db
+    redis = update_track_is_available.redis
+
+    have_lock = False
+    update_lock = redis.lock(
+        UPDATE_TRACK_IS_AVAILABLE_LOCK,
+        timeout=DEFAULT_LOCK_TIMEOUT_SECONDS,
+    )
+
+    try:
+        have_lock = update_lock.acquire(blocking=False)
+        if have_lock:
+            redis.set(
+                "update_track_is_available:start",
```

Comment:
i think we normally log durations so we can see a history of it. is there a better reason to use redis for this?

Reply:
we could either parse through logs, or check the redis key if we need to. I figure this was an easier way to check when the last job ran!

Reply:
ok 👍 it might good to add some logs too and expose job status in the verbose health check

Reply:
exposed the job statuses in the health check. I think the current logs in the task file + task logs are sufficient for what we need, in terms of logs 
```
{"levelno": 20, "level": "INFO", "msg": "Scheduler: Sending due task update_track_is_available (update_track_is_available)", "timestamp": "2022-06-06 21:31:57,034", "pathname": "/usr/lib/python3.9/site-packages/celery/beat.py", "funcName": "apply_entry", "lineno": 240, "service": "beat"}
{"levelno": 20, "level": "INFO", "msg": "Received task: update_track_is_available[2fb77cde-b70d-4a43-b8fe-95685f0e0938]  ", "timestamp": "2022-06-06 21:31:57,036", "pathname": "/usr/lib/python3.9/site-packages/celery/worker/strategy.py", "funcName": "task_message_handler", "lineno": 157, "service": "worker"}
{"levelno": 20, "level": "INFO", "msg": "Task update_track_is_available[2fb77cde-b70d-4a43-b8fe-95685f0e0938] succeeded in 0.029346704948693514s: None", "timestamp": "2022-06-06 21:31:57,066", "pathname": "/usr/lib/python3.9/site-packages/celery/app/trace.py", "funcName": "info", "lineno": 124, "service": "worker", "data": {"id": "2fb77cde-b70d-4a43-b8fe-95685f0e0938", "name": "update_track_is_available", "return_value": "None", "runtime": 0.029346704948693514}}
```